tesseract_urdf/src/urdf_parser.cpp

```diff
@@ -63,119 +102,108 @@ tesseract_common::StatusCode::Ptr parseURDFString(tesseract_scene_graph::SceneGr
     if (!sg->addLink(*l))
-      return std::make_shared<tesseract_common::StatusCode>(
-          URDFStatusCategory::ErrorAddingLinkToSceneGraph, status_cat, status);
+      std::throw_with_nested(std::runtime_error("URDF: Error adding link '" + l->getName() +
+                                                "' to scene graph for robot '" + robot_name + "'!"));
   }
 
   if (sg->getLinks().empty())
-    return std::make_shared<tesseract_common::StatusCode>(URDFStatusCategory::ErrorNoLinks, status_cat);
+    std::throw_with_nested(std::runtime_error("URDF: Error no links were found for robot '" + robot_name + "'!"));
 
   for (tinyxml2::XMLElement* joint = robot->FirstChildElement("joint"); joint; joint = joint->NextSiblingElement("join"
                                                                                                                  "t"))
   {
     tesseract_scene_graph::Joint::Ptr j = nullptr;
-    auto status = parse(j, joint, urdf_version);
-    if (!(*status))
-      return std::make_shared<tesseract_common::StatusCode>(
-          URDFStatusCategory::ErrorParsingLinkElement, status_cat, status);
+    try
+    {
+      j = parseJoint(joint, urdf_version);
+    }
+    catch (...)
+    {
+      std::throw_with_nested(std::runtime_error("URDF: Error parsing 'joint' element for robot '" + robot_name + "'!"));
+    }
 
     // Check if joint name is unique
     if (sg->getJoint(j->getName()) != nullptr)
-      return std::make_shared<tesseract_common::StatusCode>(
-          URDFStatusCategory::ErrorJointNamesNotUnique, status_cat, status);
+      std::throw_with_nested(std::runtime_error("URDF: Error joint name '" + j->getName() +
+                                                "' is not unique for robot '" + robot_name + "'!"));
 
     // Add joint to scene graph
     if (!sg->addJoint(*j))
-      return std::make_shared<tesseract_common::StatusCode>(
-          URDFStatusCategory::ErrorAddingJointToSceneGraph, status_cat, status);
+      std::throw_with_nested(std::runtime_error("URDF: Error adding joint '" + j->getName() +
+                                                "' to scene graph for robot '" + robot_name + "'!"));
   }
 
   if (sg->getJoints().empty())
-    return std::make_shared<tesseract_common::StatusCode>(URDFStatusCategory::ErrorNoJoints, status_cat);
+    std::throw_with_nested(std::runtime_error("URDF: Error no joints were found for robot '" + robot_name + "'!"));
 
   if (!sg->isTree())
   {
     if (sg->isAcyclic())
-      return std::make_shared<tesseract_common::StatusCode>(URDFStatusCategory::ErrorIsNotTree, status_cat);
+      std::throw_with_nested(
+          std::runtime_error("URDF: Error, is not a tree structure for robot '" + robot_name + "'!"));
 
-    return std::make_shared<tesseract_common::StatusCode>(URDFStatusCategory::ErrorIsAcyclic, status_cat);
+    std::throw_with_nested(std::runtime_error("URDF: Error, is acyclic structure for robot '" + robot_name + "'!"));
```

Comment:
Yea, I had to walk myself through this again when I came to it. The first checks to see if the scene graph is a tree. If it is not a tree then it must contains a link with two parents but this does not automatically indicate that the graph is acyclic so it performs the check. The acyclic check is not necessary because it is already bad but will just provided more details about the scene graph.

Reply:
But shouldn't the is acyclic error go inside the if and the generic tree error go after?

Reply:
I reworked it a little in the open PR but I believe it should be inside the if statement. In order for it to contain cycles it must be a graph and not a tree.  If it is not a tree then it checks if the graph contains cycle  and if so it throws an exception with a more detailed message.